### PR TITLE
Add DirectByteBuffer strategy for transport-nio

### PR DESCRIPTION
**Pull request by Tim-Brooks**
This is related to #27260. In Elasticsearch all of the messages that we
serialize to write to the network are composed of heap bytes. When you
read or write to a nio socket in java, the heap memory you passed down
must be copied to/from direct memory. The JVM internally does some
buffering of the direct memory, however it is essentially unbounded.

This commit introduces a simple mechanism of buffering and copying the
memory in transport-nio. Each network event loop is given a 64kb
DirectByteBuffer. When we go to read we use this buffer and copy the
data after the read. Additionally, when we go to write, we copy the data
to the direct memory before calling write. 64KB is chosen as this is the
default receive buffer size we use for transport-netty4
(`NETTY_RECEIVE_PREDICTOR_SIZE`).

Since we only have one buffer per thread, we could afford larger.
However, if we the buffer is large and not all of the data is flushed in
a `write` call, we will do excess copies. This is something we can
explore in the future.

### CODE REVIEW

File: libs/nio/src/main/java/org/elasticsearch/nio/SocketChannelContext.java [hunk 1]
```diff
@@ -234,49 +234,107 @@ protected boolean closeNow() {
...
-            return (int) rawChannel.write(buffers);
-        } catch (IOException e) {
-            closeNow = true;
-            throw e;
+        ByteBuffer ioBuffer = getSelector().getIoBuffer();
+
+        boolean continueFlush = buffers.length != 0;
+        int bytesFlushed = 0;
+        int lastBufferCopied = 0;
+        while (continueFlush) {
```
Comment (original-brownbear) [inline libs/nio/src/main/java/org/elasticsearch/nio/SocketChannelContext.java]: This loop looks like we're turning a non-blocking write into a blocking one. If a channel is not writable for some period of time (maybe become we're passing a lot of data in `buffers`) won't this approach block the IO thread?

File: libs/nio/src/main/java/org/elasticsearch/nio/SocketChannelContext.java [hunk 2]
```diff
@@ -234,49 +234,107 @@ protected boolean closeNow() {
...
+        }
     }
 
     protected int flushToChannel(ByteBuffer buffer) throws IOException {
+        ByteBuffer ioBuffer = getSelector().getIoBuffer();
+        copyBytes(buffer, ioBuffer);
+        ioBuffer.flip();
         try {
-            return rawChannel.write(buffer);
+            return rawChannel.write(ioBuffer);
```
Comment (original-brownbear) [inline libs/nio/src/main/java/org/elasticsearch/nio/SocketChannelContext.java]: Shouldn't we move the position on `buffer` back again by the number of writes not written if the write doesn't succeed in writing all of `ioBuffer`, i.e. when `ioBuffer.remaining() < rawChannel.write(ioBuffer)`?

File: libs/nio/src/main/java/org/elasticsearch/nio/SocketChannelContext.java (hunk 1 above)
Comment (tbrooks8) [reply libs/nio/src/main/java/org/elasticsearch/nio/SocketChannelContext.java]: Good catch. That should be checking that greater than 0 was flushed. I'll fix it.

Comment (tbrooks8) [reply libs/nio/src/main/java/org/elasticsearch/nio/SocketChannelContext.java]: Good catch. That should be checking that greater than 0 was flushed. I'll fix it.

Comment (tbrooks8) [reply libs/nio/src/main/java/org/elasticsearch/nio/SocketChannelContext.java]: Actually no. If we have any remaining data in the io buffer after flush we stop looping.

File: libs/nio/src/main/java/org/elasticsearch/nio/SocketChannelContext.java (hunk 2 above)
Comment (tbrooks8) [reply libs/nio/src/main/java/org/elasticsearch/nio/SocketChannelContext.java]: This is an issue. The passed buffer needs to reflect partial flushes.

**Comment by tbrooks8** [comment]
@original-brownbear - Thanks for the review. I have added a number of tests. And fixed the issue you identified.

File: libs/nio/src/main/java/org/elasticsearch/nio/SocketChannelContext.java (hunk 1 above)
Comment (original-brownbear) [reply libs/nio/src/main/java/org/elasticsearch/nio/SocketChannelContext.java]: Right my bad sorry ... misread the last line in the loop.

**Review by original-brownbear — APPROVED** [review]
@tbrooks8 thanks LGTM :)

**Comment by tbrooks8** [comment]
run packaging sample tests

**Comment by tbrooks8** [comment]
@elasticmachine run sample packaging tests

**Pull request merged**